rust/src/storage/s3/store.rs

```diff
@@ -0,0 +1,658 @@
+            } else {
+                break Ok(());
+            }
+
+            // check if more objects exists
+            continuation_token = self.unpack_next_continuation_token(&list_obj_resp);
+            if continuation_token.is_none() {
+                break Ok(());
+            }
+        }
+    }
+
+    // Get the object identifier/keys
+    fn unpack_object_identifier(
+        &self,
+        list_obj_resp: &ListObjectsV2Output,
+    ) -> Option<Vec<ObjectIdentifier>> {
+        if let Some(objects) = &list_obj_resp.contents {
+            let keys = objects
+                .into_iter()
+                .filter_map(|obj| obj.key.clone())
+                .map(|key| ObjectIdentifier {
+                    key,
+                    ..Default::default()
+                })
+                .collect();
+            Some(keys)
+        } else {
+            None
+        }
+    }
+
+    // Delete the objects of the given bucket
+    async fn delete_objects(
+        &self,
+        bucket: &str,
+        identifiers: Vec<ObjectIdentifier>,
+    ) -> Result<DeleteObjectsOutput, RusotoError<DeleteObjectsError>> {
+        let req = DeleteObjectsRequest {
+            bucket: bucket.to_string(),
+            delete: Delete {
+                objects: identifiers,
+                ..Default::default()
+            },
+            ..Default::default()
+        };
+
+        self.s3_client.delete_objects(req).await
+    }
+
+    /// Return all objects keys for the given bucket.
+    async fn list_objects(
+        // the response returns 1000 keys max.
+        // https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/S3.html#listObjectsV2-property
+        &self,
+        bucket: &str,
+        continuation_token: Option<String>,
+    ) -> Result<ListObjectsV2Output, RusotoError<ListObjectsV2Error>> {
+        // list all objects
+        let req = ListObjectsV2Request {
+            bucket: bucket.to_string(),
+            continuation_token,
+            // Minio is not limited to 1000
+            max_keys: Some(1000),
+            ..Default::default()
+        };
+
+        self.s3_client.list_objects_v2(req).await
+    }
+
+    // Unpack the next_continuation_token of the ListObjectsV2Output response
+    fn unpack_next_continuation_token(
+        &self,
+        list_obj_resp: &ListObjectsV2Output,
+    ) -> Option<String> {
+        // https://docs.aws.amazon.com/AmazonS3/latest/dev/ListingObjectKeysUsingJava.html
+        if let Some(is_truncated) = list_obj_resp.is_truncated {
+            if is_truncated {
+                list_obj_resp.next_continuation_token.clone()
+            } else {
+                None
+            }
+        } else {
+            None
+        }
+    }
+}
+
+// experimental stream implementation of the list object endpoint
+struct ListObjectIdentifierFuture(
+    Pin<
+        Box<
+            dyn Future<Output = Result<(Option<Vec<ObjectIdentifier>>, Option<String>), ()>> + Send,
```

Comment:
Some explanations about this type would be nice: what do `Option<Vec<ObjectIdentifier>>` and `Option<String>` represent? It might even be worth making this tuple a named struct (?).

Reply:
That is a good idea. I will change that